contrib/nitrpg/src/web.nit

```diff
@@ -0,0 +1,163 @@
+			exit 1
+		end
+	end
+
+	# Return an Error reponse page.
+	fun bad_request(msg: String): HttpResponse do
+		var rsp = new HttpResponse(400)
+		var page = new NitRpgPage
+		var error = new ErrorPanel(msg)
+		page.flow_panels.add error
+		rsp.body = page.write_to_string
+		return rsp
+	end
+end
+
+# An action that require a game.
+class GameAction
+	super RpgAction
+
+	# Response page stub.
+	var page: NitRpgPage is noinit
+
+	# Target game.
+	var game: Game is noinit
+
+	redef fun answer(request, url) is abstract
+
+	# Check errors and prepare response.
+	private fun prepare_response(request: HttpRequest, url: String): HttpResponse do
+		var owner = request.param("owner")
+		var repo_name = request.param("repo")
+		if owner == null or repo_name == null then
+			var msg = "Bad request: should look like /repos/:owner/:repo."
+			return bad_request(msg)
+		end
+		var repo = new Repo(api, "{owner}/{repo_name}")
+		game = new Game(api, repo)
+		if api.was_error then
+			var msg = api.last_error.message
+			return bad_request("Repo Error: {msg}")
+		end
+		var response = new HttpResponse(200)
+		page = new NitRpgPage
+		page.side_panels.add new GameStatusPanel(game)
+		page.breadcrumbs = new Breadcrumbs
+		page.breadcrumbs.add_link(game.url, game.name)
+		return response
+	end
+end
+
+# Repo overview page.
+class RepoHome
+	super GameAction
+
+	redef fun answer(request, url) do
+		var rsp = prepare_response(request, url)
+		page.side_panels.add new ShortListPlayersPanel(game)
+		page.flow_panels.add new PodiumPanel(game)
+		rsp.body = page.write_to_string
+		return rsp
+	end
+end
+
+# Repo players list.
+class ListPlayers
+	super GameAction
+
+	redef fun answer(request, url) do
+		var rsp = prepare_response(request, url)
+		page.breadcrumbs.add_link(game.url / "players", "players")
+		page.flow_panels.add new ListPlayersPanel(game)
+		rsp.body = page.write_to_string
+		return rsp
+	end
+end
+
+# Player details page.
+class PlayerHome
+	super GameAction
+
+	redef fun answer(request, url) do
+		var rsp = prepare_response(request, url)
+		var name = request.param("player")
+		if name == null then
+			var msg = "Bad request: should look like /:owner/:repo/:players/:name."
+			return bad_request(msg)
+		end
+		var player = game.load_player(name)
+		if player == null then
+			return bad_request("Request Error: unknown player {name}.")
+		end
+		page.breadcrumbs.add_link(game.url / "players", "players")
+		page.breadcrumbs.add_link(player.url, name)
+		page.side_panels.clear
+		page.side_panels.add new PlayerStatusPanel(game, player)
+		page.flow_panels.add new PlayerReviewsPanel(game, player)
+		rsp.body = page.write_to_string
+		return rsp
+	end
+end
+
+if args.length != 2 then
+	print "Error: missing argument"
+	print ""
+	print "Usage:"
+	print "web <host> <port>"
+	exit 1
+end
+
+var host = args[0]
+var port = args[1]
+
+var iface = "{host}:{port}"
+var vh = new VirtualHost(iface)
+vh.routes.add new Route("/styles/", new FileServer("www/styles"))
```

Comment:
Not needed here because the FileServer render files from the current directory (where `web` is started).

But your remarks apply for all links found in templates, so I will do it :)